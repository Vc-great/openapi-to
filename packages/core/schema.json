{
  "$schema": "http://json-schema.org/draft-07/schema",
  "$id": "@openapi-to/define-config",
  "title": "JSON schema for @openapi-to/core define-config",
  "type": "object",
  "required": [],
  "properties": {
    "root": {
      "type": "string",
      "description": "Project root directory. Can be an absolute path, or a path relative from the location of the config file itself.",
      "default": "process.cwd()"
    },
    "input": {
      "type": "object",
      "description": "",
      "properties": {
        "path": {
          "type": "string",
          "description": "Path to be used as the input. Can be an absolute path, or a path relative from the defined root option."
        }
      }
    },
    "output": {
      "type": "object",
      "description": "",
      "properties": {
        "path": {
          "type": "string",
          "description": "Path to be used to export all generated files. Can be an absolute path, or a path relative based of the defined root option."
        },
        "clean": {
          "type": "boolean",
          "description": "Remove previous generated files and folders."
        },
        "write": {
          "type": "boolean",
          "description": "Enabled or disable the writing to the filesystem. This is being used for the playground.",
          "default": "true"
        }
      }
    },
    "hooks": {
      "type": "object",
      "description": "Hooks that can be called when a specific action is done in openapi-to. ",
      "properties": {
        "done": {
          "oneOf": [
            {
              "type": "string",
              "description": "Hook that will be called at the end of all executions. Useful for running Prettier or ESLint to use your own linting structure."
            },
            {
              "type": "array",
              "items": {
                "type": "string"
              },
              "description": "Hook that will be called at the end of all executions. Useful for running Prettier or ESLint to use your own linting structure."
            }
          ]
        }
      }
    },
    "plugins": {
      "oneOf": [
        {
          "type": "array",
          "items": {
            "type": "array",
            "prefixItems": [
              {
                "type": "string"
              }
            ],
            "items": {
              "type": "object"
            }
          }
        },
        {
          "type": "object"
        }
      ],
      "description": "Plugins"
    }
  }
}
