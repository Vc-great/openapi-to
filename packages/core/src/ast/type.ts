import type { JSDocTagStructure } from "ts-morph";
import type {
  AmbientableNodeStructure,
  ExportableNodeStructure,
  JSDocableNodeStructure,
  Structure,
} from "ts-morph";
import type {
  BindingNamedNodeStructure,
  ExclamationTokenableNodeStructure,
  KindedStructure,
  OptionalKind,
  StructureKind,
  TypedNodeStructure,
  VariableDeclarationKind,
  WriterFunction,
} from "ts-morph";

export interface VariableStatementStructureOfObject
  extends Structure,
    VariableStatementSpecificStructure,
    JSDocableNodeStructure,
    AmbientableNodeStructure,
    ExportableNodeStructure {}

interface VariableStatementSpecificStructure
  extends KindedStructure<StructureKind.VariableStatement> {
  declarationKind?: VariableDeclarationKind;
  declarations: OptionalKind<VariableDeclarationStructure>[];
}

export interface VariableDeclarationStructure
  extends Structure,
    VariableDeclarationSpecificStructure,
    BindingNamedNodeStructure,
    InitializerExpressionableNodeStructure,
    TypedNodeStructure,
    ExclamationTokenableNodeStructure {}

interface VariableDeclarationSpecificStructure
  extends KindedStructure<StructureKind.VariableDeclaration> {}

export interface InitializerExpressionableNodeStructure {
  initializer?: string | WriterFunction | ObjectStructure[];
}

export interface ObjectStructure {
  key: string;
  value: string;
  docs: JSDocStructure[];
}

export interface JSDocStructure {
  /**
   * The description of the JS doc.
   * @remarks To force this to be multi-line, add a newline to the front of the string.
   */
  description?: string;
  /** JS doc tags (ex. `&#64;param value - Some description.`). */
  tags?: OptionalKind<JSDocTagStructure>[];
}

export type JSDoc = JSDocStructure[];
