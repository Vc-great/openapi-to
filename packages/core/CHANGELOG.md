# @openapi-to/core

## 3.0.1

### Patch Changes

- fix bug

## 3.0.0

### Major Changes

- fix bug
- fix: bug
- update
- refactor

## 3.0.0-alpha.5

### Major Changes

- fix: bug

## 3.0.0-alpha.4

### Major Changes

- fix bug

## 3.0.0-alpha.3

### Major Changes

- refactor

## 3.0.0-alpha.2

### Major Changes

- fix

## 2.4.0-alpha.1

### Patch Changes

- fix bug

## 2.4.0-alpha.0

### Minor Changes

- update all

## 2.3.0

### Minor Changes

- generate swr

## 2.2.0

### Minor Changes

- fix bug

## 2.1.0

### Minor Changes

- optimize request type zod

## 2.0.0

### Major Changes

- [`e519466`](https://github.com/Vc-great/openapi-to/commit/e5194667c7416e817a498d592c357a7ae9c05f22) Thanks [@Vc-great](https://github.com/Vc-great)! - v2

- [`a39a900`](https://github.com/Vc-great/openapi-to/commit/a39a9002dda434d8a65768f55c69875ed8ad1eea) - fix bug

- 2.0

- [`367be25`](https://github.com/Vc-great/openapi-to/commit/367be252aa434487c09c4566e77792839867b509) - fix bug

- [`7fd9e58`](https://github.com/Vc-great/openapi-to/commit/7fd9e58417ef5563dedf945fbb030b70b8b09bd6) - rc

## 2.0.0-rc.5

### Major Changes

- rc

## 2.0.0-alpha.4

### Major Changes

- fix bug

## 2.0.0-alpha.3

### Major Changes

- fix bug

## 2.0.0-alpha.2

### Major Changes

- v2
