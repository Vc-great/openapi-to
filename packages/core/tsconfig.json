{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "baseUrl": ".",
    "types": ["vitest/globals", "bun-types"]
  },
  "include": ["src/**/*.ts", "./package.json", "vitest.config.ts", "tsup.config.ts", "../config-ts/reset.d.ts"]
}
