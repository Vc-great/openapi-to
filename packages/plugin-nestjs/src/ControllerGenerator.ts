import { NUMBER_ENUM } from "@openapi-to/core";
import { URLPath } from "@openapi-to/core/utils";

import _ from "lodash";
import { Scope } from "ts-morph";

import { HttpStatusObject } from "./http-status.ts";
import { NestjsGenerator } from "./NestjsGenerator.ts";

import type { PluginContext } from "@openapi-to/core";
import type { ParameterObject } from "oas/types";
import type { OpenAPIV3 } from "openapi-types";
import type { ClassDeclarationStructure } from "ts-morph";
import type { ParameterDeclarationStructure } from "ts-morph";
import type { DecoratorStructure } from "ts-morph";
import type {
  ImportDeclarationStructure,
  MethodDeclarationStructure,
  OptionalKind,
} from "ts-morph";
import type { Config } from "./types.ts";

type ImportStatementsOmitKind = Omit<ImportDeclarationStructure, "kind">;

export class ControllerGenerator extends NestjsGenerator {
  private nestjsCommonImports: Set<string> = new Set<string>();
  private swaggerImports: Set<string> = new Set<string>();

  constructor(config: Config) {
    super(config);
  }

  get getResponseStatusCodes() {
    return this.openapi.response?.getResponseStatusCodes;
  }

  get hasAuth() {
    return _.has(this.oas.api, "components.securitySchemes.http");
  }

  get isArrayOfRequestBody(): boolean {
    const requestBodyObject = this.openapi.requestBody?.requestBodyObject;
    return _.get(requestBodyObject, "schemaObject.type") === "array";
  }

  get isArrayORBaseType() {
    const requestBodyObject = this.openapi.requestBody?.requestBodyObject;
    const requestBodyType =
      requestBodyObject?.refName ||
      _.get(requestBodyObject, "schemaObject.$ref") ||
      _.get(requestBodyObject, "schemaObject.type");

    return (
      this.isArrayOfRequestBody ||
      ["string", "number", "boolean"].includes(
        _.isArray(requestBodyType) ? "" : (requestBodyType ?? ""),
      )
    );
  }

  get hasPermissions(): boolean {
    return !!this.oas.getExtension("x-permissions", this.operation);
  }

  build(context: PluginContext): void {
    _.forEach(this.openapi.pathGroupByTag, (pathGroup, tag) => {
      const methodsStatements = _.chain(pathGroup)
        .map(({ path, method, tag }) => {
          this.operation = this.openapi.setCurrentOperation(path, method, tag);

          this.setNestjsCommonImports();
          this.setSwaggerImports();
          return this.generatorMethod();
        })
        .value();

      this.ast.createSourceFile(this.controllerFilePath, {
        statements: [
          ...this.generateImport(),
          this.generatorClass(methodsStatements),
        ],
      });
    });
  }

  getParseArrayPipe(): string | undefined {
    const mediaTypeObject =
      this.openapi.requestBody?.getRequestBodySchemaOfApplicationJson;

    const type = _.get(mediaTypeObject, "schema.items.type", "");
    //todo ("array" | NonArraySchemaObjectType)[]
    if (_.isArray(type)) {
      return undefined;
    }

    const isBaseTypeOfBodyArrayItems = !["object", "array"].includes(type);

    const baseType = isBaseTypeOfBodyArrayItems
      ? NUMBER_ENUM.includes(type)
        ? "Number"
        : type
      : type;

    return baseType
      ? `new ParseArrayPipe({ items: ${_.upperFirst(baseType)} })`
      : undefined;
  }

  setNestjsCommonImports(): void {
    this.nestjsCommonImports.add(_.upperFirst(this.operation?.method));

    const hasPathNumber = _.some(
      this.openapi.parameter?.parametersOfPath,
      (item) => NUMBER_ENUM.includes(_.get(item, "schema.type", "")),
    );

    _.chain([] as string[])
      .concat(this.openapi.parameter?.hasQueryParameters ? "Query" : [])
      .concat(this.openapi.parameter?.hasPathParameters ? "Param" : [])
      .concat(this.openapi.requestBody?.hasRequestBody ? "Body" : [])
      .concat(hasPathNumber ? ["Param", "ParseIntPipe"] : [])
      .concat(this.getParseArrayPipe() ? ["ParseArrayPipe"] : [])
      .forEach((x) => this.nestjsCommonImports.add(x))
      .value();
  }

  setSwaggerImports(): void {
    _.chain([] as string[])
      .concat(this.openapi.parameter?.hasPathParameters ? "ApiParam" : [])
      .concat(this.isArrayORBaseType ? "ApiBody" : [])
      .concat(this.hasAuth ? "ApiBearerAuth" : [])
      .forEach((x) => {
        this.swaggerImports.add(x);
      })
      .value();
  }

  generateImport(): Array<ImportDeclarationStructure> {
    const nestjsCommon: ImportStatementsOmitKind = {
      namedImports: [
        "Controller",
        "HttpStatus",
        "HttpCode",
        ...this.nestjsCommonImports,
      ],
      moduleSpecifier: "@nestjs/common",
    };

    const nestjsSwagger: ImportStatementsOmitKind = {
      namedImports: ["ApiOperation", "ApiResponse", ...this.swaggerImports],
      moduleSpecifier: "@nestjs/swagger",
    };
    //todo 筛选
    const auth: ImportStatementsOmitKind = {
      namedImports: ["Permissions"], //"Public"
      moduleSpecifier: "@/common/decorators/auth.decorator",
    };

    const ApiTag: ImportStatementsOmitKind = {
      namedImports: ["ApiTag"],
      moduleSpecifier: "@/common/swagger",
    };

    const service: ImportStatementsOmitKind = {
      namedImports: [this.serviceClassName],
      moduleSpecifier: this.serviceModuleSpecifier,
    };

    const statements = _.chain([] as Array<ImportStatementsOmitKind>)
      .push(nestjsCommon)
      .push(nestjsSwagger)
      .push(auth)
      .push(ApiTag)
      .push(service)
      .concat(this.domainImportStatements)
      .filter(Boolean)
      .value();

    return this.ast.generateImportStatements(statements);
  }

  generatorClass(
    methodsStatements: OptionalKind<MethodDeclarationStructure>[],
  ): ClassDeclarationStructure {
    const statements = {
      name: this.controllerClassName,
      decorators: this.generatorClassDecorators(),
      isExported: true,
      docs: [],
      methods: methodsStatements,
      ctors: [
        {
          parameters: [
            {
              isReadonly: true,
              scope: Scope.Private,
              name: _.lowerFirst(this.serviceClassName),
              type: this.serviceClassName,
            },
          ],
        },
      ],
    };
    return this.ast.generateClassStatements(statements);
  }

  generatorClassDecorators(): OptionalKind<DecoratorStructure>[] {
    const apiTags = {
      name: "ApiTag",
      arguments: [
        `{
          name: '${this.openapi.currentTagName}',
          description: '${this.openapi.currentTagMetadata?.description}',
        }`,
      ],
    };

    const controller = {
      name: "Controller",
      arguments: [`'${this.openapi.currentTagName}'`],
    };

    const apiBearerAuth = {
      name: "ApiBearerAuth",
      arguments: [],
    };

    return _.chain([] as OptionalKind<DecoratorStructure>[])
      .push(apiTags)
      .push(controller)
      .concat(this.hasAuth ? apiBearerAuth : [])
      .filter(Boolean)
      .value();
  }

  generatorMethodDecorators(): OptionalKind<DecoratorStructure>[] {
    const httpCode: string = _.get(
      HttpStatusObject,
      this.openapi.response?.successCode || "200",
    );

    const regex = `/${this.openapi.currentTagName}`;

    const methodPath = new URLPath(
      this.openapi.operation?.path.replace(regex, "") || "",
    ).toURLPath;

    const methodDecorator = {
      name:
        _.upperFirst(this.openapi.operation?.method) +
        (methodPath ? "('" : "(") +
        methodPath +
        (methodPath ? "')" : ")"),
    };

    const httpCodeDecorator = {
      name: "HttpCode",
      arguments: [`HttpStatus.${httpCode}`],
    };

    const ApiOperationDecorator: OptionalKind<DecoratorStructure> = {
      name: "ApiOperation",
      arguments: [
        "{" +
          _.chain([] as string[])
            .concat(
              this.operation?.getSummary()
                ? `summary:'${this.operation?.getSummary()}'`
                : [],
            )
            .concat(
              this.operation?.getDescription()
                ? `description:'${this.operation?.getDescription()}'`
                : [],
            )
            .join()
            .value() +
          "}",
      ],
    };

    const responseType = this.openapi.response?.successResponse?.refName;

    const successResponseJSONSchema = this.openapi.response?.successResponse;

    const ApiResponseDecorator: OptionalKind<DecoratorStructure> = {
      name: "ApiResponse",
      arguments: [
        "{" +
          _.chain([] as string[])
            .push(`status: HttpStatus.${httpCode}`)
            .concat(
              successResponseJSONSchema?.description
                ? `description:${JSON.stringify(successResponseJSONSchema?.description)}`
                : [],
            )
            .concat(responseType ? responseType : [])
            .concat(
              successResponseJSONSchema?.type === "array"
                ? ` isArray:true`
                : [],
            )
            .filter(Boolean)
            .join(",")
            .value() +
          "}",
      ],
    };

    const requestBodyObject = this.openapi.requestBody?.requestBodyObject;

    const requestBodyType =
      requestBodyObject?.refName ||
      _.get(requestBodyObject, "schemaObject.$ref") ||
      _.get(requestBodyObject, "schemaObject.type");

    const apiBodyDecorator: OptionalKind<DecoratorStructure> = {
      name: "ApiBody",
      arguments: [
        JSON.stringify({
          required: _.get(
            this.operation,
            "schema.requestBody.required",
            undefined,
          ),
          type: requestBodyType || "any",
          isArray: this.isArrayOfRequestBody || undefined,
          description: requestBodyObject?.description,
          examples: requestBodyObject?.examples,
          example: requestBodyObject?.example,
        }),
      ],
    };

    const apiParam: OptionalKind<DecoratorStructure>[] = _.chain(
      this.openapi.parameter?.parameters,
    )
      .filter(["in", "path"])
      .map((item) => {
        return {
          name: "ApiParam",
          arguments: [
            `{
              name:"${_.camelCase(item.name)}",
              description:"${item.description}"
            }`,
          ],
        };
      })
      .value();

    const permissions = this.oas.getExtension(
      "x-permissions",
      this.operation,
    ) as string[] | undefined;

    const PermissionsDecorator: OptionalKind<DecoratorStructure> = {
      name: "Permissions",
      arguments: permissions
        ? permissions.map((item) => JSON.stringify(item))
        : [],
    };

    return _.chain([] as OptionalKind<DecoratorStructure>[])
      .concat(ApiOperationDecorator)
      .concat(ApiResponseDecorator)
      .concat(apiParam)
      .concat(this.isArrayORBaseType ? apiBodyDecorator : [])
      .concat(methodDecorator)
      .concat(this.hasPermissions ? PermissionsDecorator : [])
      .concat(httpCodeDecorator)
      .filter(Boolean)
      .value();
  }

  /**
   * @example
   * ```
   * @Query() query: QueryUserDto
   * @Param('id', ParseIntPipe)id: number
   * @Body() data: UpdateUserDto
   * ```
   */
  generatorMethodParameters(): OptionalKind<ParameterDeclarationStructure>[] {
    const queryParameters = {
      ...this.queryParameters,
      decorators: [
        {
          name: "Query",
          arguments: [],
        },
      ],
    };

    const bodyParameters = {
      ...this.bodyParameters,
      decorators: [
        {
          name: "Body",
          arguments: [this.getParseArrayPipe()].filter(Boolean),
        },
      ],
    };

    const pathParameters = _.chain(this.openapi.parameter?.parameters)
      .filter(["in", "path"])
      .map((item: OpenAPIV3.ParameterObject) => {
        //todo other type
        const pathType =
          (item.schema && "type" in item.schema && item.schema.type) ??
          "string";
        return {
          name: _.camelCase(item.name),
          type: NUMBER_ENUM.includes(pathType || "") ? "number" : pathType,
          decorators: [
            {
              name: "Param",
              arguments: NUMBER_ENUM.includes(pathType || "")
                ? [JSON.stringify(_.camelCase(item.name)), "ParseIntPipe"]
                : [],
            },
          ],
        };
      })
      .value() as unknown as OptionalKind<ParameterDeclarationStructure>[];

    return _.chain([] as OptionalKind<ParameterDeclarationStructure>[])
      .concat(this.openapi.parameter?.hasPathParameters ? pathParameters : [])
      .concat(this.openapi.parameter?.hasQueryParameters ? queryParameters : [])
      .concat(this.openapi.requestBody?.hasRequestBody ? bodyParameters : [])
      .filter(Boolean)
      .value();
  }

  generatorMethodBody(): string {
    const pathParameters = _.chain(this.openapi.parameter?.parameters)
      .filter(["in", "path"])
      .map((item: ParameterObject) => _.camelCase(item.name))
      .value() as unknown as string[];

    const args = _.chain([] as string[])
      .concat(this.openapi.parameter?.hasPathParameters ? pathParameters : [])
      .concat(this.openapi.parameter?.hasQueryParameters ? "query" : [])
      .concat(this.openapi.requestBody?.hasRequestBody ? "data" : [])
      .join()
      .value();

    return `return await this.${_.lowerFirst(this.serviceClassName)}.${this.methodName}(${args})`;
  }

  generatorMethod(): MethodDeclarationStructure {
    const statement = {
      name: this.methodName,
      decorators: this.generatorMethodDecorators(),
      parameters: this.generatorMethodParameters(),
      returnType: this.generatorReturnType(),
      docs: [],
      statements: this.generatorMethodBody(),
      isAsync: true,
    };

    return this.ast.generateMethodStatements(statement);
  }
}
