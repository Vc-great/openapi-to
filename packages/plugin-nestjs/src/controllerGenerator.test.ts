import path from "node:path";
import { dirname } from "node:path";
import { fileURLToPath } from "node:url";

import { AST, OpenAPI } from "@openapi-to/core";

import _ from "lodash";
import Oas from "oas";
import { describe } from "vitest";

import petStore from "../mock/petstore.json";
import { ControllerGenerator } from "./ControllerGenerator.ts";

import type { OpenapiToSingleConfig } from "@openapi-to/core";
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

describe("RequestGenerator", async () => {
  const openapiToSingleConfig: OpenapiToSingleConfig = {
    name: "",
    root: "",
    input: {
      path: "",
    },
    output: {
      dir: path.resolve(__dirname, "../mock"),
    },
    plugins: [],
  };

  const context = {
    output: "",
  };

  test("controllerGenerator", () => {
    const ast = new AST();
    // @ts-expect-error Not a canonical document
    const oas = new Oas(petStore);
    const openapi = new OpenAPI({}, oas);
    const pluginConfig = {};

    const controllerGenerator = new ControllerGenerator({
      oas,
      ast,
      openapi,
      pluginConfig,
      openapiToSingleConfig,
    });
    controllerGenerator.build(context);
    const text = _.chain(ast.sourceFile)
      .map((sourceFile) => sourceFile.getFullText())
      .join("\n")
      .value();

    expect(text).toMatchSnapshot();
  });
});
