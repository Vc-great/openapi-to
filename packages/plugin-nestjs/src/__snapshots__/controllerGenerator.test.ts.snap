// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RequestGenerator > controllerGenerator 1`] = `
"import { Controller, HttpStatus, HttpCode, Get, Query, Post, Body, Put, Delete, ParseArrayPipe, Param, ParseIntPipe } from "@nestjs/common";
import { ApiOperation, ApiResponse, ApiBearerAuth, ApiBody, ApiParam } from "@nestjs/swagger";
import { Permissions } from "@/common/decorators/auth.decorator";
import { ApiTag } from "@/common/swagger";
import { PetService } from "./pet.service";
import { FindAllQueryDto } from "./domain/findAll-query.dto";
import { FindAllUserVo } from "./domain/FindAllUser.vo";
import { testBodyDto } from "./domain/testBodyDto.dto";
import { testDto2 } from "./domain/testDto2.vo";
import { test32145 } from "./domain/test32145.vo";
import { CreateUpdateUserDto } from "./domain/CreateUpdateUserDto.dto";
import { TestidgetQueryDto } from "./domain/testidget-query.dto";
import { ApiResponse } from "./domain/ApiResponse.vo";
import { Pet } from "./domain/Pet.dto";
import { FindPetsByStatusQueryDto } from "./domain/findPetsByStatus-query.dto";
import { FindPetsByTagsQueryDto } from "./domain/findPetsByTags-query.dto";

@ApiTag({
    name: 'pet',
    description: 'Everything about your Pets',
})
@Controller('pet')
@ApiBearerAuth()
export class PetController {
    constructor(private readonly petService: PetService) {
    }

    @ApiOperation({ summary: 'get接口', description: '获取用户列表' })
    @ApiResponse({ status: HttpStatus.OK, description: "success111111", FindAllUserVo })
    @Get('/test')
    @HttpCode(HttpStatus.OK)
    async findAll(@Query() query: FindAllQueryDto): Promise<FindAllUserVo> {
        return await this.petService.findAll(query)
    }

    @ApiOperation({ summary: 'summary' })
    @ApiResponse({ status: HttpStatus.OK, description: "OK", testDto2 })
    @Post('/test')
    @Permissions("admin")
    @HttpCode(HttpStatus.OK)
    async test(@Body() data: testBodyDto): Promise<testDto2> {
        return await this.petService.test(data)
    }

    @ApiOperation({ summary: 'summary' })
    @ApiResponse({ status: HttpStatus.OK, description: "OK", testDto2 })
    @Put('/test')
    @HttpCode(HttpStatus.OK)
    async testput(@Body() data: testBodyDto): Promise<testDto2> {
        return await this.petService.testput(data)
    }

    @ApiOperation({ summary: 'summary' })
    @ApiResponse({ status: HttpStatus.OK, description: "OK", test32145 })
    @ApiBody({ "required": true, "type": "array", "isArray": true, "description": "ids" })
    @Delete('/test')
    @HttpCode(HttpStatus.OK)
    async testdelete(@Body(new ParseArrayPipe({ items: Number })) data: number[]): Promise<test32145> {
        return await this.petService.testdelete(data)
    }

    @ApiOperation({ summary: 'summary' })
    @ApiResponse({ status: HttpStatus.OK, description: "OK", testDto2 })
    @ApiBody({ "required": true, "type": "array", "isArray": true, "description": "批量创建用户dto", "examples": { "a": { "summary": "概要", "description": "事例说明", "value": { "a": 1, "b": "2" } } } })
    @Post('/test/batch-create')
    @Permissions("admin")
    @HttpCode(HttpStatus.OK)
    async batchCreate(@Body() data: CreateUpdateUserDto[]): Promise<testDto2> {
        return await this.petService.batchCreate(data)
    }

    @ApiOperation({ summary: 'summary' })
    @ApiResponse({ status: HttpStatus.OK, description: "OK", testDto2 })
    @ApiParam({
        name: "testId",
        description: "test-id"
    })
    @Get('/test/:testId')
    @HttpCode(HttpStatus.OK)
    async testidget(@Param("testId", ParseIntPipe) testId: number, @Query() query: TestidgetQueryDto): Promise<testDto2> {
        return await this.petService.testidget(testId, query)
    }

    @ApiOperation({ summary: 'uploads an image', description: 'pet' })
    @ApiResponse({ status: HttpStatus.OK, description: "successful operation", ApiResponse })
    @ApiParam({
        name: "petId",
        description: "ID of pet to update"
    })
    @Post('/:petId/uploadImage')
    @HttpCode(HttpStatus.OK)
    async uploadFile(@Param("petId", ParseIntPipe) petId: number, @Body() data: any): Promise<ApiResponse> {
        return await this.petService.uploadFile(petId, data)
    }

    @ApiOperation({ summary: 'Add a new pet to the store' })
    @ApiResponse({ status: HttpStatus.OK })
    @Post()
    @HttpCode(HttpStatus.OK)
    async addPet(@Body() data: Pet): Promise<void> {
        return await this.petService.addPet(data)
    }

    @ApiOperation({ summary: 'Update an existing pet' })
    @ApiResponse({ status: HttpStatus.OK })
    @Put()
    @HttpCode(HttpStatus.OK)
    async updatePet(@Body() data: Pet): Promise<void> {
        return await this.petService.updatePet(data)
    }

    @ApiOperation({ summary: 'Finds Pets by status', description: 'Multiple status values can be provided with comma separated strings' })
    @ApiResponse({ status: HttpStatus.OK, description: "successful operation", isArray: true })
    @Get('/findByStatus')
    @HttpCode(HttpStatus.OK)
    async findPetsByStatus(@Query() query: FindPetsByStatusQueryDto): Promise<Pet[]> {
        return await this.petService.findPetsByStatus(query)
    }

    @ApiOperation({ summary: 'Finds Pets by tags', description: 'Multiple tags can be provided with comma separated strings. Use tag1, tag2, tag3 for testing.' })
    @ApiResponse({ status: HttpStatus.OK, description: "successful operation", isArray: true })
    @Get('/findByTags')
    @HttpCode(HttpStatus.OK)
    async findPetsByTags(@Query() query: FindPetsByTagsQueryDto): Promise<Pet[]> {
        return await this.petService.findPetsByTags(query)
    }

    @ApiOperation({ summary: 'Find pet by ID', description: 'Returns a single pet' })
    @ApiResponse({ status: HttpStatus.OK, description: "successful operation", Pet })
    @ApiParam({
        name: "petId",
        description: "ID of pet to return"
    })
    @Get('/:petId')
    @HttpCode(HttpStatus.OK)
    async getPetById(@Param("petId", ParseIntPipe) petId: number): Promise<Pet> {
        return await this.petService.getPetById(petId)
    }

    @ApiOperation({ summary: 'Updates a pet in the store with form data' })
    @ApiResponse({ status: HttpStatus.OK })
    @ApiParam({
        name: "petId",
        description: "ID of pet that needs to be updated"
    })
    @Post('/:petId')
    @HttpCode(HttpStatus.OK)
    async updatePetWithForm(@Param("petId", ParseIntPipe) petId: number, @Body() data: any): Promise<void> {
        return await this.petService.updatePetWithForm(petId, data)
    }

    @ApiOperation({ summary: 'Deletes a pet' })
    @ApiResponse({ status: HttpStatus.OK })
    @ApiParam({
        name: "petId",
        description: "Pet id to delete"
    })
    @Delete('/:petId')
    @HttpCode(HttpStatus.OK)
    async deletePet(@Param("petId", ParseIntPipe) petId: number): Promise<void> {
        return await this.petService.deletePet(petId)
    }
}

import { Controller, HttpStatus, HttpCode, Get, Query, Post, Body, Put, Delete, ParseArrayPipe, Param, ParseIntPipe } from "@nestjs/common";
import { ApiOperation, ApiResponse, ApiBearerAuth, ApiBody, ApiParam } from "@nestjs/swagger";
import { Permissions } from "@/common/decorators/auth.decorator";
import { ApiTag } from "@/common/swagger";
import { StoreService } from "./store.service";
import { Order } from "./domain/Order.dto";

@ApiTag({
    name: 'store',
    description: 'Access to Petstore orders',
})
@Controller('store')
@ApiBearerAuth()
export class StoreController {
    constructor(private readonly storeService: StoreService) {
    }

    @ApiOperation({ summary: 'Place an order for a pet' })
    @ApiResponse({ status: HttpStatus.OK, description: "successful operation", Order })
    @Post('/order')
    @HttpCode(HttpStatus.OK)
    async placeOrder(@Body() data: Order): Promise<Order> {
        return await this.storeService.placeOrder(data)
    }

    @ApiOperation({ summary: 'Find purchase order by ID', description: 'For valid response try integer IDs with value >= 1 and <= 10. Other values will generated exceptions' })
    @ApiResponse({ status: HttpStatus.OK, description: "successful operation", Order })
    @ApiParam({
        name: "orderId",
        description: "ID of pet that needs to be fetched"
    })
    @Get('/order/:orderId')
    @HttpCode(HttpStatus.OK)
    async getOrderById(@Param("orderId", ParseIntPipe) orderId: number): Promise<Order> {
        return await this.storeService.getOrderById(orderId)
    }

    @ApiOperation({ summary: 'Delete purchase order by ID', description: 'For valid response try integer IDs with positive integer value. Negative or non-integer values will generate API errors' })
    @ApiResponse({ status: HttpStatus.OK })
    @ApiParam({
        name: "orderId",
        description: "ID of the order that needs to be deleted"
    })
    @Delete('/order/:orderId')
    @HttpCode(HttpStatus.OK)
    async deleteOrder(@Param("orderId", ParseIntPipe) orderId: number): Promise<void> {
        return await this.storeService.deleteOrder(orderId)
    }

    @ApiOperation({ summary: 'Returns pet inventories by status', description: 'Returns a map of status codes to quantities' })
    @ApiResponse({ status: HttpStatus.OK, description: "successful operation" })
    @Get('/inventory')
    @HttpCode(HttpStatus.OK)
    async getInventory(): Promise<void> {
        return await this.storeService.getInventory()
    }
}

import { Controller, HttpStatus, HttpCode, Get, Query, Post, Body, Put, Delete, ParseArrayPipe, Param, ParseIntPipe } from "@nestjs/common";
import { ApiOperation, ApiResponse, ApiBearerAuth, ApiBody, ApiParam } from "@nestjs/swagger";
import { Permissions } from "@/common/decorators/auth.decorator";
import { ApiTag } from "@/common/swagger";
import { UserService } from "./user.service";
import { UserArray } from "./domain/UserArray.dto";
import { User } from "./domain/User.vo";
import { LoginUserQueryDto } from "./domain/loginUser-query.dto";

@ApiTag({
    name: 'user',
    description: 'Operations about user',
})
@Controller('user')
@ApiBearerAuth()
export class UserController {
    constructor(private readonly userService: UserService) {
    }

    @ApiOperation({ summary: 'Creates list of users with given input array' })
    @ApiResponse({ status: HttpStatus.OK })
    @ApiBody({ "type": "array", "isArray": true, "description": "List of user object" })
    @Post('/createWithArray')
    @HttpCode(HttpStatus.OK)
    async createUsersWithArrayInput(@Body() data: UserArray[]): Promise<void> {
        return await this.userService.createUsersWithArrayInput(data)
    }

    @ApiOperation({ summary: 'Creates list of users with given input array' })
    @ApiResponse({ status: HttpStatus.OK })
    @ApiBody({ "type": "array", "isArray": true, "description": "List of user object" })
    @Post('/createWithList')
    @HttpCode(HttpStatus.OK)
    async createUsersWithListInput(@Body() data: UserArray[]): Promise<void> {
        return await this.userService.createUsersWithListInput(data)
    }

    @ApiOperation({ summary: 'Get user by user name' })
    @ApiResponse({ status: HttpStatus.OK, description: "successful operation", User })
    @ApiParam({
        name: "username",
        description: "The name that needs to be fetched. Use user1 for testing. "
    })
    @Get('/:username')
    @HttpCode(HttpStatus.OK)
    async getUserByName(@Param() username: string): Promise<User> {
        return await this.userService.getUserByName(username)
    }

    @ApiOperation({ summary: 'Updated user', description: 'This can only be done by the logged in user.' })
    @ApiResponse({ status: HttpStatus.OK })
    @ApiParam({
        name: "username",
        description: "name that need to be updated"
    })
    @Put('/:username')
    @HttpCode(HttpStatus.OK)
    async updateUser(@Param() username: string, @Body() data: User): Promise<void> {
        return await this.userService.updateUser(username, data)
    }

    @ApiOperation({ summary: 'Delete user', description: 'This can only be done by the logged in user.' })
    @ApiResponse({ status: HttpStatus.OK })
    @ApiParam({
        name: "username",
        description: "The name that needs to be deleted"
    })
    @Delete('/:username')
    @HttpCode(HttpStatus.OK)
    async deleteUser(@Param() username: string): Promise<void> {
        return await this.userService.deleteUser(username)
    }

    @ApiOperation({ summary: 'Logs user into the system' })
    @ApiResponse({ status: HttpStatus.OK, description: "successful operation" })
    @Get('/login')
    @HttpCode(HttpStatus.OK)
    async loginUser(@Query() query: LoginUserQueryDto): Promise<void> {
        return await this.userService.loginUser(query)
    }

    @ApiOperation({ summary: 'Logs out current logged in user session' })
    @ApiResponse({ status: HttpStatus.OK })
    @Get('/logout')
    @HttpCode(HttpStatus.OK)
    async logoutUser(): Promise<void> {
        return await this.userService.logoutUser()
    }

    @ApiOperation({ summary: 'Create user', description: 'This can only be done by the logged in user.' })
    @ApiResponse({ status: HttpStatus.OK })
    @Post()
    @HttpCode(HttpStatus.OK)
    async createUser(@Body() data: User): Promise<void> {
        return await this.userService.createUser(data)
    }
}
"
`;
